### PR TITLE
Custom sort function for InitContainersStatuses

### PR DESCRIPTION
**Pull request by lukaszo**
Order in init containers matters. Statues shoudln't be sorted by name.

### CODE REVIEW

**Comment by lukaszo** [comment]
@smarterclayton can you please take a look at this?

**Comment by lukaszo** [comment]
The order is wrong because of this sort: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L546

**Comment by smarterclayton** [comment]
Hrm, that sort is probably wrong then.

**Comment by smarterclayton** [comment]
We should probably fix the sort of statuses - it should always be in order of the init container initial positions.

**Comment by lukaszo** [comment]
Ok, I will take a look  at it

**Comment by smarterclayton** [comment]
Bumping to p1 since this results in inconsistent client experience with init containers.

**Comment by lukaszo** [comment]
Ok, there was one more sort call. I changed it so InitContainerStatuses is not sorted now. With this changes kubectl shows correct status now.

Maybe in long term InitContainerStatuses should have different type which strictly forbids sorting. It would require much more changes though.

**Comment by smarterclayton** [comment]
Well, the statuses still need to be sorted - they just need to be in a specific sort that is not the same as app containers.  We depend on the sort order of init container statuses in some cases.

**Comment by smarterclayton** [comment]
Basically the sort we need to use for init containers is:

"order that init containers appear in the pod"

**Comment by lukaszo** [comment]
Ah, I thought that it will be in correct order now. My bad. 
Ok, so I see two solutions here:
* write custom sort function which will use data from pod.Spec.InitContainers to get correct order
* add `index` field to ContainerStatus struct and create Interface to sort by this index

This `index` field would not make any sense for pod containers so I think first option is better. What do you think?  Or do you see better approach?

**Comment by smarterclayton** [comment]
I think the first approach makes sense - you should be able to swap the two positions as you walk the status array (find the first name, swap it into the first position, find the second name, swap it into the second position) in classic bubble sort.

**Comment by lukaszo** [comment]
It's getting more complicated. In normalizeStatus from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L507  I have only access to api.PodStatus and it's not enough to sort InitContainers.
Any idea how to solve it? Passing pod.Spec  through all functions sound like a bad idea :/

**Comment by smarterclayton** [comment]
How many places call normalizeStatus?  The only use is see has the pod spec
available.

On Jun 6, 2016, at 6:34 PM, Łukasz Oleś <notifications@github.com> wrote:

It's getting more complicated. In normalizeStatus from
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L507
I have only access to api.PodStatus and it's not enough to sort
InitContainers.
Any idea how to solve it? Passing pod.Spec through all functions sound like
a bad idea :/

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/26829#issuecomment-224109634>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_py0f7t5JM9r_dKf3Nz3UzVBeIbykks5qJKCNgaJpZM4IuAeQ>
.

**Comment by lukaszo** [comment]
It's called from `updateStatusInternal` and from `needsReconcile` in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go `needReconcile` is called from syncBatch

**Comment by lukaszo** [comment]
If passing pod to normalizeStatus is ok for you then there is no problem.

**Comment by smarterclayton** [comment]
It looks like both cases you can pass pod spec in, that is ok with me.

**Comment by lukaszo** [comment]
I added unit test for sort function. Seems to work. Are there any other cases to check?

File: pkg/kubelet/types/types.go [hunk 1]
```diff
@@ -68,6 +68,21 @@ func (s SortedContainerStatuses) Less(i, j int) bool {
 	return s[i].Name < s[j].Name
 }
 
+// Sorts InitContainerStatuses
```
Comment (smarterclayton) [inline pkg/kubelet/types/types.go]: Minor thing, but we try to stay consistent to Godoc style which is `SortInitContainerStatuses ensures that statuses are in the order that their init container appears in the pod spec."  I.e. one sentence minimum, starts with the name of the function.

**Comment by smarterclayton** [comment]
One minor comment, if you can fix that and squash I'll apply the label.  Thanks!

**Comment by lukaszo** [comment]
Ok,  copy/paste of your text

**Comment by smarterclayton** [comment]
Thanks, squash

**Comment by lukaszo** [comment]
done

File: pkg/kubelet/types/types.go [hunk 2]
```diff
@@ -68,6 +68,22 @@ func (s SortedContainerStatuses) Less(i, j int) bool {
 	return s[i].Name < s[j].Name
 }
 
+// SortInitContainerStatuses ensures that statuses are in the order that their
+// init container appears in the pod spec
+func SortInitContainerStatuses(p *api.Pod, statuses []api.ContainerStatus) {
+	containers := p.Spec.InitContainers
+	current := 0
+	for _, container := range containers {
+		for j := current; j < len(statuses); j++ {
```
Comment (smarterclayton) [inline pkg/kubelet/types/types.go]: Hrm, if someone passes a statuses where container.Name is not in statuses, will this loop forever?

Comment (lukaszo) [inline pkg/kubelet/types/types.go]: Hmm, I don't see how. Could you provide test case for unit test?

Comment (smarterclayton) [inline pkg/kubelet/types/types.go]: Never mind, it looks fine.  Just wanted to be extra sure.

**Comment by smarterclayton** [comment]
Looks like the unit test fails to build

**Comment by lukaszo** [comment]
Yes, I fixed unit tests. Now I'm trying to find the issue with GCE Node e2e tests

**Comment by smarterclayton** [comment]
Node tests were not you - was a bug today:

@k8s-bot test this issue #26431

**Comment by lukaszo** [comment]
@smarterclayton can you add label LGTM? Or do you have any more comments?

**Comment by smarterclayton** [comment]
Was waiting for the submit queue to get back to normal.  Marked now, thanks again.

**Comment by lukaszo** [comment]
@smarterclayton thank you for your help and patience  :)

**Pull request merged**